### PR TITLE
Build: Generate stable module and chunk IDs and a better chunk hash

### PR DESCRIPTION
**Pull request by blowery**
The goal of the PR is to make chunk hashes, chunk IDs and module IDs stable over time. Stable chunk hashes require stable module IDs and stable chunks IDs. Webpack uses monotonically increasing IDs by default; this results in instability at multiple levels. 

#### Module IDs
`react` may map to module ID `4` in one build and then map to module ID `5` in the next. Modules that import `react` would then change, even if they themselves did not. This, in turn, should make the chunk hash change when it should not.

This PR moves us to a stable module name -> module ID mapping, based on the md5 of the module path. This makes module IDs stable across builds.

A couple folks asked if we could we use the `NamedModulesPlugin` instead. We definitely could, though the Webpack folks are now recommending the HashedModulesPlugin for production, which is what this replicates.

#### Chunk IDs

As we add more and more chunks with `asyncRequire`, we add more and more chunks to the mix. Chunk IDs, like Module IDs, are just an increasing ID by default. This means that Chunk A may map to ID `3` in one build and chunk `5` in the next. This means that other chunks that depend on Chunk A would need to update their hash, even if nothing within the chunk changes. 

This PR moves us to a stable chunk name -> chunk ID mapping, based on the chunk's `name` property. It requires named chunks, which all of our chunks are, once `commons` is removed. 

Now when we introduce new chunks, those chunks get a new unique ID and references remain stable

#### Chunk Hashes

The default webpack chunk hash generator is pretty complicated and depends on things that are no longer necessary when we have stable module IDs and stable chunk IDs.

This PR adds a new chunk hasher that uses an MD5 hash of the content plus the hash seed to generate the hash.

### Further Explanation

The module ID generator is lifting mostly from `webpack-stable-module-id-and-hash`. It works quite well, and could be replaced with the new `HashedModuleIdPlugin` when we move to `webpack@2`. There's no way to turn off the chunk hash generator in `webpack-stable-module-id-and-hash` though, so I just pulled it. Plus we have control if we need to modify it.

The chunk ID generator is based on it, but uses the chunk name instead of
the module path as a key into the chunk ID.

The chunk hash generator is based on `webpack-chunk-hash`.  We could probably just use `webpack-chunk-hash` directly, but having the code locally will likely be useful.

This also moves the vendor dll into a commons chunk. Otherwise we can't control the module IDs generated by the DLL Reference plugin. The module IDs within the vendor DLL were fine, but the DllRefererncePlugin aliases them to a monotonically increasing ID which is not stable across builds. We need all the module and chunk IDs to be stable to make this all work. 

I also removed the commons chunk since it was now down to 4k zipped and only really needed by a few sections. This saves 4k for a bunch of section loads.

## Testing

To test this, first make a base build with `CALYPSO_ENV=production make build`. Copy off `public` to a safe place (`cp -r public base-public`). Now start fiddling with things in the source tree and npm dependencies and see how the built code changes (`diff -rqu base-public public`).

Build calypso off the commit prior to this one (not current master). Compare to this PR. Do the hashes for all of the chunks change? They should.

When changing something in `vendor`, only `vendor` should change. Try updating `react` or `moment` or `react-redux`. Expect to see a new vendor chunk and no other changes. 

When something in an asyncRequire chunk changes, only that chunk should change. Try modifying code in `reader/sidebar/expandable.jsx`. What chunks get new hashes? It should be limited to `async-load-reader-sidebar`.

When you add a new module to the build chunk, only the build chunk should change. Try importing something new from `client/boot`. What changes? It should just be the build chunk. The new module should get a new unique ID and other module IDs should not be affected.

### CODE REVIEW

**Comment by blowery** [comment]
To test:

- [ ] Apply react-redux@5 and see if we see hash movement (#11753)
- [ ] Apply the happy chat async load and verify that chunk hashes change (#11868)

File: server/bundler/webpack/stable-chunk-hash.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+		} );
+		if ( chunk.chunks.length ) {
+			collectChunkIds( chunk.chunks, chunkIds );
+		}
+	} );
+
+	return chunkIds;
+}
+
+function collectChunkIds( chunk ) {
```
Comment (blowery) [inline server/bundler/webpack/stable-chunk-hash.js]: drop this? not needed.

**Comment by blowery** [comment]
I added a bunch more explanation to the summary of the PR that should help with testing (and I hope understanding).

**Comment by ockham** [comment]
Rebased on current `master`, and did the same to #11753 so we can more easily compare their build dirs.

**Comment by ockham** [comment]
> Apply react-redux@5 and see if we see hash movement (#11753)

```
git co fix/build/chunk-hash-issues
make clean
CALYPSO_ENV=production make build
cp -r public before
git co upgrade/react-redux-take-four
git rebase fix/build/chunk-hash-issues
make clean
CALYPSO_ENV=production make build
cp -r public after
diff -rqu before after
```

```
Only in before: async-load-components-post-schedule.4d8ed09eb72085d8724c.js
Only in before: async-load-components-post-schedule.4d8ed09eb72085d8724c.m.js
Only in after: async-load-components-post-schedule.7c30c240386c43af350e.js
Only in after: async-load-components-post-schedule.7c30c240386c43af350e.m.js
Only in after: devdocs.317af305d33c1970aae5.js
Only in after: devdocs.317af305d33c1970aae5.m.js
Only in before: devdocs.ec5db93c9543ae13ffa1.js
Only in before: devdocs.ec5db93c9543ae13ffa1.m.js
Only in before: manifest.ab44162de13bf77b908f.js
Only in before: manifest.ab44162de13bf77b908f.m.js
Only in after: manifest.d2fdc777ace1b52a483a.js
Only in after: manifest.d2fdc777ace1b52a483a.m.js
Files before/style-debug.css and after/style-debug.css differ
Files before/style-debug.css.map and after/style-debug.css.map differ
Files before/style-rtl.css and after/style-rtl.css differ
Files before/style.css and after/style.css differ
Only in after: theme.0bcced3ed0f92569f992.js
Only in after: theme.0bcced3ed0f92569f992.m.js
Only in before: theme.d246ec722ca5ea132e8c.js
Only in before: theme.d246ec722ca5ea132e8c.m.js
Only in before: vendor.116112c83596de24b5be.js
Only in before: vendor.116112c83596de24b5be.m.js
Only in after: vendor.eb8eb98035cb2cb45201.js
Only in after: vendor.eb8eb98035cb2cb45201.m.js
```

**Comment by ockham** [comment]
And, as per @blowery's request, `diff -rqu master before`:

https://gist.github.com/ockham/e5175929d87194b3779aba02457facbd

**Comment by samouri** [comment]
Tested this out locally in a docker build.  All LGTM 👍.

File: server/bundler/plugin.js [hunk 2]
```diff
@@ -29,11 +29,11 @@ ChunkFileNames.prototype.apply = function( compiler ) {
 					this.indent( [
 						"script.onerror = script.onload = script.onreadystatechange = null;",
 						"delete installedChunks[ chunkId ];",
-						"window.__chunkErrors[ " + JSON.stringify( chunkMaps.name ) + "[chunkId]||chunkId ]=new Error();",
+						"window.__chunkErrors[ " + JSON.stringify( chunkMaps.name, null, '  ' ) + "[chunkId]||chunkId ]=new Error();",
```
Comment (samouri) [inline server/bundler/plugin.js]: what does the null do?

File: webpack-dll.config.js [hunk 3]
```diff
@@ -3,6 +3,7 @@
  */
 const path = require( 'path' );
 const webpack = require( 'webpack' );
+const WebpackStableModuleId = require( './server/bundler/webpack/stable-module-id' );
```
Comment (blowery) [inline webpack-dll.config.js]: we don't even need this file now

File: server/bundler/plugin.js [hunk 4]
```diff
@@ -18,7 +18,7 @@ ChunkFileNames.prototype.apply = function( compiler ) {
 					"// start chunk loading",
 					"installedChunks[ chunkId ] = [ callback ];",
 					"window.__chunkErrors = window.__chunkErrors || {};",
-					"window.__chunkErrors[ " + JSON.stringify( chunkMaps.name ) + "[chunkId]||chunkId ]=null;",
+					"window.__chunkErrors[ " + JSON.stringify( chunkMaps.name, null, '  ' ) + "[chunkId]||chunkId ]=null;",
```
Comment (blowery) [inline server/bundler/plugin.js]: fix these to not use spaces for formatting

File: server/bundler/plugin.js (hunk 2 above)
Comment (blowery) [inline server/bundler/plugin.js]: the `null` is the custom formatter. you have to pass it to get at the third arg, which is the string to use when pretty printing. Dropped it all and went back to plain JSON printing. This was really just for debugging. Makes the JSON easier to read.

File: webpack-dll.config.js (hunk 3 above)
Comment (blowery) [inline webpack-dll.config.js]: dropped

File: server/bundler/plugin.js (hunk 4 above)
Comment (blowery) [inline server/bundler/plugin.js]: done

**Pull request merged**